src/Concordium/Client/Types/Contract/Schema.hs

```diff
@@ -716,9 +792,19 @@ putLenWithSizeLen sl len = case sl of
 
 -- * Map *
 
-getMapOfWithSizeLen :: Ord k => SizeLength -> S.Get k -> S.Get v -> S.Get (Map k v)
-getMapOfWithSizeLen sl gt gv = S.label "Map" $ Map.fromList <$> getListOfWithSizeLen sl (S.getTwoOf gt gv)
+-- |Get a map with a specified size length. Fails if the predicate is false when applied to the list of tuples in the map.
+getMapOfWithSizeLenAndPred :: (Ord k) => ([(k,v)] -> Bool) -> SizeLength -> S.Get k -> S.Get v -> S.Get (Map k v)
+getMapOfWithSizeLenAndPred p sl gt gv = do
+  ls <- getListOfWithSizeLen sl (S.getTwoOf gt gv)
+  if p ls
+  then S.label "Map" $ pure $ Map.fromList ls
```

Comment:
I am not sure what the `label` there achieves. This will never fail so I don't think the label will ever be printed.

IN general, this is also not the way to do things. You have in essence made all the `getMapOfWithSizeLen` slower. In practice this won't matter **in the client for the specific case we have**, since the amount of data is sufficiently small, but we do need to be careful when doing things like this.

Reply:
Label is there since I assumed that the code I modified was sound by induction. Should I remove it?

What is making this so terribly slow, and what is the better way of accomplishing this?  I hope and pray that GHC is capable of inferring that

`getMapOfWithSizeLenAndPred (const True) sl gt gv` \equiv `\sl gt gv -> S.label "Map" $ Map.fromList <$> getListOfWithSizeLen sl (S.getTwoOf gt gv)`

Maybe there is something I misunderstand :-)

Reply:
It is not that it is **terribly** slow, and as I wrote in my comment, in practice it most likely will not be observable.

But it is slower than it was.
- First a list is constructed
- Then a list is traversed, applying a function to it.
- Then the list is converted to a map.

You might say "laziness makes this OK" and it is true, **it might**.

I would not use this function to implement `getMapOfWithSizeLen`. But again, since the number of elements here will be at most a few thousand it will not be noticeable and this is not worth more of a discussion.

My general point is that abstractions have a cost and we should be mindful of the cost at every level and not make things needlessly slower.